internal/worker/run.go

```diff
@@ -243,17 +345,40 @@ func (w *Worker) runTaskLogic(ctx context.Context, syslog, log logr.Logger, slot
 	defer cleanupContainer(log, proxyContainer)
 
 	startCh := make(chan error, 1)
-	nextCh := startCh
 	exitCh := make(chan error, 1)
+	leaves := []<-chan error{startCh}
 	go watchContainer(ctx, syslog, proxyContainer, exitCh)
 
 	// Configure containers
 	for idx := range payload.Steps {
 		stepCh := make(chan error, 1)
-		go w.runStep(ctx, log, sandbox, claim, proxyContainer, idx, &payload, []<-chan error{nextCh}, stepCh)
-		nextCh = stepCh
+		go w.runStep(ctx, log, sandbox, claim, proxyContainer, idx, &payload, leaves, stepCh)
+		leaves = []<-chan error{stepCh}
+	}
+
+	// Configure artifact uploads
+	artifactDependencies := leaves
+	leaves = make([]<-chan error, len(payload.Artifacts))
+	for idx := range payload.Artifacts {
+		artifact := &payload.Artifacts[idx]
+		stepCh := make(chan error, 1)
+		go w.uploadArtifact(ctx, log, claim, proxyContainer, artifact, artifactDependencies, stepCh)
+		leaves[idx] = stepCh
 	}
 
+	// Gather all leaves
+	nextCh := make(chan error, 1)
```

Comment:
it's fine, i trust this all works, but all this manual juggling of "leaves" muddles my brain

the language implies it's a graph, but my understanding is that right now it's all coded to be a simple chain of tasks, so there are assumptions in the algorithms - I think.

if this ever got more complicated (e.g. with actual parallelisation or whatever), I'd want to see these channels abstracted slightly somehow to more intuitively describe the fact that there are things depending on prior tasks, or something.

Reply:
> if this ever got more complicated (e.g. with actual parallelisation or whatever), I'd want to see these channels abstracted slightly somehow to more intuitively describe the fact that there are things depending on prior tasks, or something.
I'm a little confused here... the channels _are_ this abstraction... I'm not sure I can go more abstract...

Reply:
I think, having now seen the psuedo YAML spec for a future graph impl, this will end up less confusing for other reasons:

I think you calling it "leaves" has confused me, but effectively in the new world order you'll end up with named steps, effectively a map of names to "completion channels", and the "dependsOn" array will look up the completion channels to depend on for any given step, with the default being the completion channel of the previous step (just like the current implementation).